### PR TITLE
Conf Store - Basic Framework

### PR DESCRIPTION
**Pull request by suryakumar1024**
Conf Store changes:
Updated conf_store file and added conf class which works on top of conf_store
fixed numeric key value handling.
fixed file load.
updated test cased for both ConfStore and Conf.

### CODE REVIEW

File: py-utils/src/utils/kv_store/kv_store_collection.py [hunk 1]
```diff
@@ -77,7 +87,7 @@ def dump(self, data) -> None:
 
 
 class IniKvStore(KvStore):
-    """  Represents a YAML File Store """ 
+    """  Represents a YAML File Store """
```
Comment (ujjwalpl) [inline py-utils/src/utils/kv_store/kv_store_collection.py:90]: INI File Store

File: py-utils/src/utils/kv_store/kvstore.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+#!/bin/python3
```
Comment (ujjwalpl) [inline py-utils/src/utils/kv_store/kvstore.py:1]: Remove this file. This is not needed any more. There is already a kv_store.py.

File: py-utils/src/utils/kv_store/pillar.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
+#!/bin/python3
```
Comment (ujjwalpl) [inline py-utils/src/utils/kv_store/pillar.py:1]: Remove this. This is already a part of kv_store_collection.py

File: py-utils/test/conf_store_example.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import sys
```
Comment (ujjwalpl) [inline py-utils/test/conf_store_example.py:1]: Remove this file. test should only contain test cases not the example code. 
OR 
convert this into a test using test format.

File: py-utils/test/conf_store_example.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+sys.path.insert(1, '../')
+
+from src.utils.conf_store import ConfStore, KvStorageFactory, KvStore
```
Comment (ujjwalpl) [inline py-utils/test/conf_store_example.py:4]: from cortx.utils.conf_store

File: py-utils/test/test_conf_store/test_conf_store.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        load_config('test_local', 'json:///tmp/file1.json')
+        result_data = conf_store.get('test_local', 'bridge.name', default_val=None)
+        self.assertEqual(result_data, 'Homebridge')
+
+    def test_conf_store_get_wrong_key(self):
+        """
+        Test by trying to get the wrong key from the loaded config
+        """
+        load_config('new_local', 'json:///tmp/file1.json')
+        result_data = conf_store.get('test_local', 'bridge.no_name_field', default_val=None)
```
Comment (ujjwalpl) [inline py-utils/test/test_conf_store/test_conf_store.py:86]: Follow 80 char limit

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/test/test_conf_store/test_conf.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+sample_config = Json(file_path).load()
+
+
+def setup_and_generate_sample_files():
+    """ This function will generate all required types of file """
+    with open(r'/tmp/config.json', 'w+') as file:
+        json.dump(sample_config, file, indent=2)
+
+
+conf = Conf()
```
Comment (ujjwalpl) [inline py-utils/test/test_conf_store/test_conf.py:40]: No need to instantiate Conf class. You can directly use it.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/conf_store/conf_cache.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        self._dirty = False
+
+    def _refresh_keys(self, data, pkey: str = None):
+        if type(data) == list:
+            for i in range(len(data)):
+                self._keys.append("%s[%d]" %(pkey, i))
+        elif type(data) == dict: 
+            for key in data.keys():
+                nkey = key if pkey is None else "%s.%s" %(pkey, key)
+                if type(data[key]) == str or type(data[key]) == int: self._keys.append(nkey)
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_cache.py:59]: Follow indentation

File: py-utils/src/utils/conf_store/conf_store.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+
+import errno
+
+from cortx.utils.conf_store.error import ConfStoreError
+from cortx.utils.conf_store.conf_cache import ConfCache
+from cortx.utils.kv_store.kv_store import KvStoreFactory
+
+CONF_CACHE = ConfCache
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_store.py:24]: No need to declare this. Pls remove

File: py-utils/src/utils/conf_store/conf_cache.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        self._iter += 1
+        return key
+
+    def _get(self, data: dict, key: str):
+        """ Obtain value for the given key """
+        k = key.split('.', 1)
+        if k[0] not in data.keys(): return None
+        return self._get(data[k[0]], k[1]) if len(k) > 1 else data[k[0]]
+
+    def get(self, key: str):
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_cache.py:81]: We can add key=None condition here. 

``` 
def get(self, key: str = None):
    if key is None: return self._data
    return self._get(self._data, key)
```

File: py-utils/src/utils/conf_store/conf_store.py [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+        default_val
+                Default Value
+
+        Return Value:
+                Return type will be dict or string based of key
+        """
+        if index not in self._cache.keys():
+            raise ConfStoreError(errno.EINVAL, "config index %s is not loaded",
+                                 index)
+        if key is not None:
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_store.py:76]: This condition can be delegated down to conf_cache.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/kv_store/kvstore.py (hunk 2 above)
Comment (suryakumar1024) [reply py-utils/src/utils/kv_store/kvstore.py:1]: Removed from kv_store in current patch. I have already removed but still i can see this file.

File: py-utils/src/utils/kv_store/pillar.py (hunk 3 above)
Comment (suryakumar1024) [reply py-utils/src/utils/kv_store/pillar.py:1]: Removed from kv_store in current patch.

File: py-utils/test/conf_store_example.py (hunk 4 above)
Comment (suryakumar1024) [reply py-utils/test/conf_store_example.py:1]: Already testcases created for this package. So removed in current patch

File: py-utils/test/conf_store_example.py (hunk 5 above)
Comment (suryakumar1024) [reply py-utils/test/conf_store_example.py:4]: This file was removed.

File: py-utils/test/test_conf_store/test_conf.py (hunk 7 above)
Comment (suryakumar1024) [reply py-utils/test/test_conf_store/test_conf.py:40]: Updated in current patch

File: py-utils/src/utils/conf_store/conf_cache.py (hunk 8 above)
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_cache.py]: Update in current patch.

File: py-utils/test/test_conf_store/test_conf_store.py (hunk 6 above)
Comment (suryakumar1024) [reply py-utils/test/test_conf_store/test_conf_store.py]: Update in current patch

File: py-utils/src/utils/kv_store/kv_store_collection.py (hunk 1 above)
Comment (suryakumar1024) [reply py-utils/src/utils/kv_store/kv_store_collection.py]: Update in current patch

File: py-utils/src/utils/kv_store/kv_store_collection.py [hunk 12]
```diff
@@ -178,11 +188,11 @@ def get(self, key):
 
         except Exception as ex:
             raise KvError(errno.ENOENT, f"Cant get data for %s. %s.", \
-                key, ex)
+                          key, ex)
```
Comment (ujjwalpl) [inline py-utils/src/utils/kv_store/kv_store_collection.py:191]: Use single tab only.

File: py-utils/src/utils/kv_store/kv_store_collection.py [hunk 13]
```diff
@@ -178,11 +188,11 @@ def get(self, key):
 
         except Exception as ex:
             raise KvError(errno.ENOENT, f"Cant get data for %s. %s.", \
-                key, ex)
+                          key, ex)
 
         if res is None:
             raise KvError(errno.ENOENT, f"Cant get data for %s. %s." \
-                f"Key not present")
+                                        f"Key not present")
```
Comment (ujjwalpl) [inline py-utils/src/utils/kv_store/kv_store_collection.py:195]: Use single tab

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/conf_store/conf_store.py (hunk 11 above)
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_store.py]: Updated. Move this condition to Conf_cache

File: py-utils/src/utils/conf_store/conf_cache.py (hunk 10 above)
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_cache.py]: Updated same with the code

File: py-utils/src/utils/kv_store/kv_store_collection.py (hunk 12 above)
Comment (suryakumar1024) [reply py-utils/src/utils/kv_store/kv_store_collection.py]: Updated in current patch

File: py-utils/src/utils/kv_store/kv_store_collection.py (hunk 13 above)
Comment (suryakumar1024) [reply py-utils/src/utils/kv_store/kv_store_collection.py:194]: updated in current patch

File: py-utils/src/utils/conf_store/conf_store.py (hunk 9 above)
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_store.py]: Updated in current patch

**Comment by suryakumar1024** [comment]
I have raised new PR as requested by ujjwal. new PR link https://github.com/Seagate/cortx-utils/pull/131

**Pull request closed without merging**